pkg/controllers/telemetry/controller.go

```diff
@@ -0,0 +1,270 @@
+		CPUUsage:       cpuUsage,
+		MemoryUsage:    memoryUsage,
+		CPURequest:     cpuRequest,
+		MemoryRequest:  memoryRequest,
+		CPUCapacity:    cpuCapacity,
+		MemoryCapacity: memoryCapacity,
+		InstanceTypes:  instanceTypes,
+	}
+
+	data, err := json.Marshal(clusterResourcesInfo)
+	if err != nil {
+		log.FromContext(ctx).Error(err, "failed to marshal cluster resource info")
+		return
+	}
+
+	push(ctx, fmt.Sprintf(defaultTelemetryEndpoint+resourcesInfoPath, clusterID), data)
+}
+
+type InterruptionEvent struct {
+	CloudProvider           string    `json:"cloudProvider"`
+	Timestamp               time.Time `json:"timestamp"`
+	Region                  string    `json:"region"`
+	Zone                    string    `json:"zone"`
+	InstanceType            string    `json:"instanceType"`
+	InstanceCreateTimestamp time.Time `json:"instanceCreateTimestamp"`
+}
+
+func PushInterruptionEvent(ctx context.Context, clusterID string, node *corev1.Node) {
+	data, err := json.Marshal(InterruptionEvent{
+		CloudProvider:           cloudprovider.CloudProviderName,
+		Timestamp:               time.Now(),
+		Region:                  node.Labels[corev1.LabelZoneRegion],
+		Zone:                    node.Labels[corev1.LabelZoneFailureDomain],
+		InstanceType:            node.Labels[corev1.LabelInstanceTypeStable],
+		InstanceCreateTimestamp: node.CreationTimestamp.Time,
+	})
+	if err != nil {
+		log.FromContext(ctx).Error(err, "failed to marshal spot interruption event")
+		return
+	}
+
+	push(ctx, fmt.Sprintf(defaultTelemetryEndpoint+interruptionEventPath, clusterID), data)
+}
+
+func getClusterResourcesCapacity(ctx context.Context, kubeclient client.Client) (cpuCapacity, memoryCapacity int64, instanceTypes []string) {
+	nodeList := &corev1.NodeList{}
+
+	if err := kubeclient.List(ctx, nodeList); err != nil {
+		log.FromContext(ctx).Error(err, "failed to list nodes")
+		return
+	}
+
+	nodes := make([]corev1.Node, 0)
+	for ni := range nodeList.Items {
+		if _, ok := nodeList.Items[ni].Labels[karpenterInitializedKey]; !ok {
+			nodes = append(nodes, nodeList.Items[ni])
+		}
+	}
+
+	instanceTypeM := make(map[string]struct{})
+	for ni := range nodes {
+		cpuCapacity += nodes[ni].Status.Capacity.Cpu().MilliValue()
+		memoryCapacity += nodes[ni].Status.Capacity.Memory().Value()
+
+		if instanceType, ok := nodes[ni].Labels[corev1.LabelInstanceTypeStable]; ok && instanceType != "" {
+			instanceTypeM[instanceType] = struct{}{}
+		}
+	}
+
+	instanceTypes = lo.Keys(instanceTypeM)
+	return
+}
+
+func getClusterResourcesRequest(ctx context.Context, kubeclient client.Client) (cpuRequest, memoryRequest int64) {
+	podList := &corev1.PodList{}
+	if err := kubeclient.List(ctx, podList); err != nil {
+		log.FromContext(ctx).Error(err, "failed to list pods")
+		return
+	}
+
+	for pi := range podList.Items {
+		if podList.Items[pi].Status.Phase == corev1.PodRunning {
+			cpuRequest += podList.Items[pi].Spec.Containers[0].Resources.Requests.Cpu().MilliValue()
```

Comment:
Do we only need to count the first container's request?

Reply:
nope, for the resources involved in the calculation see

https://github.com/kubernetes-sigs/karpenter/blob/3e0c51ac84f2b3f6663e0a04403f5871b2074f61/pkg/utils/resources/resources.go#L27-L36

Reply:
Thanks, but I've already used `RequestsForPods`

https://github.com/cloudpilot-ai/karpenter-provider-alibabacloud/blob/de2cadf4185400ac733f535c3dec46bb523dd21c/pkg/controllers/telemetry/controller.go#L211-#L228